File: relo2france-schengen-tracker/mobile/android/app/src/main/java/com/relo2france/schengen/network/ApiClient.kt
/**
 * ApiClient.kt
 *
 * REST API client for WordPress backend.
 * Handles authentication, requests, and error handling.
 *
 * @package R2F_Schengen_Tracker
 * @since   1.0.0
 */

package com.relo2france.schengen.network

import com.relo2france.schengen.data.*
import com.relo2france.schengen.util.SecureStorage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

class ApiClient private constructor() {

    companion object {
        @Volatile
        private var INSTANCE: ApiClient? = null

        fun getInstance(): ApiClient {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: ApiClient().also { INSTANCE = it }
            }
        }

        private const val DEFAULT_BASE_URL = "https://relo2france.com/wp-json/r2f-schengen/v1"
    }

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        encodeDefaults = true
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val baseUrl: String
        get() = SecureStorage.load(SecureStorage.Keys.API_BASE_URL) ?: DEFAULT_BASE_URL

    private val authToken: String?
        get() = SecureStorage.getAuthToken()

    // MARK: - Authentication

    /**
     * Login with WordPress credentials
     */
    suspend fun login(username: String, password: String): Result<AuthResponse> {
        return post("/auth/login", mapOf("username" to username, "password" to password))
    }

    /**
     * Logout and invalidate token
     */
    suspend fun logout(): Result<Unit> {
        val result = post<Map<String, Boolean>>("/auth/logout", emptyMap<String, String>())
        SecureStorage.clearAuth()
        return result.map { }
    }

    // MARK: - App Status

    /**
     * Get current app status
     */
    suspend fun getAppStatus(): Result<AppStatusResponse> {
        return get("/app/status")
    }

    // MARK: - Passport Control

    /**
     * Get passport control data for border display
     */
    suspend fun getPassportControl(): Result<PassportControlData> {
        return get("/passport-control")
    }

    // MARK: - Trips

    /**
     * Get all trips
     */
    suspend fun getTrips(): Result<List<Trip>> {
        return get("/trips")
    }

    /**
     * Create a new trip
     */
    suspend fun createTrip(trip: Trip): Result<Trip> {
        return post("/trips", trip)
    }

    /**
     * Update an existing trip
     */
    suspend fun updateTrip(id: Int, trip: Trip): Result<Trip> {
        return put("/trips/$id", trip)
    }

    /**
     * Delete a trip
     */
    suspend fun deleteTrip(id: Int): Result<Unit> {
        return delete("/trips/$id")
    }

    // MARK: - Locations

    /**
     * Submit location readings in batch
     */
    suspend fun submitLocations(locations: List<LocationReading>): Result<List<SyncResult>> {
        return post("/locations/batch", mapOf("locations" to locations))
    }

    // MARK: - Sync

    /**
     * Perform full sync
     */
    suspend fun sync(request: SyncRequest): Result<SyncResponse> {
        return post("/sync", request)
    }

    /**
     * Get changes since last sync
     */
    suspend fun getChanges(since: String?): Result<List<ServerChange>> {
        val params = since?.let { "?since=$it" } ?: ""
        return get("/changes$params")
    }

    // MARK: - Device Registration

    /**
     * Register device for push notifications
     */
    suspend fun registerDevice(registration: DeviceRegistration): Result<Map<String, Boolean>> {
        return post("/device/register", registration)
    }

    /**
     * Unregister device
     */
    suspend fun unregisterDevice(deviceId: String): Result<Map<String, Boolean>> {
        return post("/device/unregister", mapOf("device_id" to deviceId))
    }

    // MARK: - HTTP Methods

    private suspend inline fun <reified T> get(endpoint: String): Result<T> {
        return request("GET", endpoint, null)
    }

    private suspend inline fun <reified T, reified B> post(endpoint: String, body: B): Result<T> {
        val jsonBody = json.encodeToString(body)
        return request("POST", endpoint, jsonBody)
    }

    private suspend inline fun <reified T, reified B> put(endpoint: String, body: B): Result<T> {
        val jsonBody = json.encodeToString(body)
        return request("PUT", endpoint, jsonBody)
    }

    private suspend inline fun <reified T> delete(endpoint: String): Result<T> {
        return request("DELETE", endpoint, null)
    }

    private suspend inline fun <reified T> request(
        method: String,
        endpoint: String,
        body: String?
    ): Result<T> = withContext(Dispatchers.IO) {
        try {
            val url = "$baseUrl$endpoint"

            val requestBuilder = Request.Builder()
                .url(url)
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")

            // Add auth header if token exists
            authToken?.let {
                requestBuilder.header("Authorization", "Bearer $it")
            }

            // Set method and body
            val requestBody = body?.toRequestBody("application/json".toMediaType())
            when (method) {
                "GET" -> requestBuilder.get()
                "POST" -> requestBuilder.post(requestBody ?: "".toRequestBody())
                "PUT" -> requestBuilder.put(requestBody ?: "".toRequestBody())
                "DELETE" -> requestBuilder.delete(requestBody)
            }

            val response = client.newCall(requestBuilder.build()).execute()
            val responseBody = response.body?.string() ?: ""

            if (response.isSuccessful) {
                if (T::class == Unit::class) {
                    @Suppress("UNCHECKED_CAST")
                    Result.success(Unit as T)
                } else {
                    val parsed = json.decodeFromString<T>(responseBody)
                    Result.success(parsed)
                }
            } else {
                val error = try {
                    json.decodeFromString<ApiError>(responseBody)
                } catch (e: Exception) {
                    ApiError("unknown", "Request failed with status ${response.code}")
                }
                Result.failure(ApiException(error.code, error.message, response.code))
            }
        } catch (e: IOException) {
            Result.failure(ApiException("network_error", "Network error: ${e.message}", 0))
        } catch (e: Exception) {
            Result.failure(ApiException("parse_error", "Failed to parse response: ${e.message}", 0))
        }
    }
}

/**
 * API exception with error details
 */
class ApiException(
    val code: String,
    override val message: String,
    val statusCode: Int
) : Exception(message)

File: relo2france-schengen-tracker/mobile/ios/SchengenTracker/Services/APIClient.swift
/**
 * APIClient.swift
 *
 * HTTP client for communicating with the WordPress REST API.
 * Handles authentication, request/response encoding, and error handling.
 *
 * @package MyTravelStatus
 * @since   1.0.0
 */

import Foundation

actor APIClient {
    static let shared = APIClient()

    // MARK: - Configuration

    private let baseURL: URL
    private var authToken: String?
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder
    private let session: URLSession

    /// App version for status checks
    private let appVersion = "1.0.0"

    // MARK: - Initialization

    private init() {
        // TODO: Configure from environment or settings
        self.baseURL = URL(string: "https://your-site.com/wp-json/r2f-schengen/v1")!

        // Configure decoder
        decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .iso8601

        // Configure encoder
        encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        encoder.dateEncodingStrategy = .iso8601

        // Configure session
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        session = URLSession(configuration: config)

        // Load saved auth token
        authToken = KeychainHelper.load(key: "authToken")
    }

    // MARK: - Authentication

    func login(email: String, password: String) async throws {
        let loginURL = baseURL.deletingLastPathComponent().appendingPathComponent("jwt-auth/v1/token")

        var request = URLRequest(url: loginURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["username": email, "password": password]
        request.httpBody = try encoder.encode(body)

        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        struct TokenResponse: Decodable {
            let token: String
        }

        let tokenResponse = try decoder.decode(TokenResponse.self, from: data)
        authToken = tokenResponse.token
        KeychainHelper.save(key: "authToken", value: tokenResponse.token)
    }

    func logout() {
        authToken = nil
        KeychainHelper.delete(key: "authToken")
    }

    var isAuthenticated: Bool {
        authToken != nil
    }

    // MARK: - App Status

    func getAppStatus() async throws -> AppStatus {
        var components = URLComponents(url: baseURL.appendingPathComponent("app/status"), resolvingAgainstBaseURL: true)!
        components.queryItems = [
            URLQueryItem(name: "version", value: appVersion),
            URLQueryItem(name: "platform", value: "ios")
        ]

        let request = URLRequest(url: components.url!)
        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        return try decoder.decode(AppStatus.self, from: data)
    }

    // MARK: - Passport Control

    func getPassportControlData() async throws -> PassportControlData {
        let request = try authenticatedRequest(path: "passport-control", method: "GET")
        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        return try decoder.decode(PassportControlData.self, from: data)
    }

    // MARK: - Trips

    func getTrips() async throws -> [Trip] {
        let request = try authenticatedRequest(path: "trips", method: "GET")
        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        return try decoder.decode([Trip].self, from: data)
    }

    func createTrip(_ trip: Trip) async throws -> Trip {
        var request = try authenticatedRequest(path: "trips", method: "POST")
        request.httpBody = try encoder.encode(trip)

        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        return try decoder.decode(Trip.self, from: data)
    }

    func updateTrip(_ trip: Trip) async throws -> Trip {
        var request = try authenticatedRequest(path: "trips/\(trip.id)", method: "PUT")
        request.httpBody = try encoder.encode(trip)

        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        return try decoder.decode(Trip.self, from: data)
    }

    func deleteTrip(id: Int) async throws {
        let request = try authenticatedRequest(path: "trips/\(id)", method: "DELETE")
        let (_, response) = try await session.data(for: request)
        try validateResponse(response)
    }

    // MARK: - Sync

    func sync(request syncRequest: SyncRequest) async throws -> SyncResponse {
        var request = try authenticatedRequest(path: "sync", method: "POST")
        request.httpBody = try encoder.encode(syncRequest)

        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        return try decoder.decode(SyncResponse.self, from: data)
    }

    func getChanges(since: Date) async throws -> [ServerChange] {
        var components = URLComponents(url: baseURL.appendingPathComponent("changes"), resolvingAgainstBaseURL: true)!

        let formatter = ISO8601DateFormatter()
        components.queryItems = [
            URLQueryItem(name: "since", value: formatter.string(from: since))
        ]

        var request = URLRequest(url: components.url!)
        try addAuthHeader(to: &request)

        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        struct ChangesResponse: Decodable {
            let changes: [ServerChange]
        }

        let changesResponse = try decoder.decode(ChangesResponse.self, from: data)
        return changesResponse.changes
    }

    // MARK: - Locations

    func uploadLocations(_ locations: [LocationReading]) async throws -> BatchUploadResult {
        var request = try authenticatedRequest(path: "locations/batch", method: "POST")

        struct BatchRequest: Encodable {
            let locations: [LocationReading]
        }

        request.httpBody = try encoder.encode(BatchRequest(locations: locations))

        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        return try decoder.decode(BatchUploadResult.self, from: data)
    }

    // MARK: - Device Registration

    func registerDevice(_ registration: DeviceRegistration) async throws {
        var request = try authenticatedRequest(path: "device/register", method: "POST")
        request.httpBody = try encoder.encode(registration)

        let (_, response) = try await session.data(for: request)
        try validateResponse(response)
    }

    func unregisterDevice(deviceId: String) async throws {
        var request = try authenticatedRequest(path: "device/unregister", method: "POST")

        struct UnregisterRequest: Encodable {
            let deviceId: String
        }

        request.httpBody = try encoder.encode(UnregisterRequest(deviceId: deviceId))

        let (_, response) = try await session.data(for: request)
        try validateResponse(response)
    }

    // MARK: - Summary

    func getSummary() async throws -> JurisdictionSummary {
        let request = try authenticatedRequest(path: "summary", method: "GET")
        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        return try decoder.decode(JurisdictionSummary.self, from: data)
    }

    // MARK: - Family

    func getFamilyMembers() async throws -> [FamilyMember] {
        let request = try authenticatedRequest(path: "family", method: "GET")
        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        return try decoder.decode([FamilyMember].self, from: data)
    }

    // MARK: - Private Helpers

    private func authenticatedRequest(path: String, method: String) throws -> URLRequest {
        guard let token = authToken else {
            throw APIError.notAuthenticated
        }

        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        return request
    }

    private func addAuthHeader(to request: inout URLRequest) throws {
        guard let token = authToken else {
            throw APIError.notAuthenticated
        }
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    }

    private func validateResponse(_ response: URLResponse) throws {
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        switch httpResponse.statusCode {
        case 200...299:
            return
        case 401:
            throw APIError.notAuthenticated
        case 403:
            throw APIError.forbidden
        case 404:
            throw APIError.notFound
        case 409:
            throw APIError.conflict
        case 500...599:
            throw APIError.serverError(httpResponse.statusCode)
        default:
            throw APIError.httpError(httpResponse.statusCode)
        }
    }
}

// MARK: - API Errors

enum APIError: Error, LocalizedError {
    case notAuthenticated
    case forbidden
    case notFound
    case conflict
    case invalidResponse
    case serverError(Int)
    case httpError(Int)

    var errorDescription: String? {
        switch self {
        case .notAuthenticated:
            return "Please log in to continue"
        case .forbidden:
            return "You don't have permission to access this resource"
        case .notFound:
            return "The requested resource was not found"
        case .conflict:
            return "There was a sync conflict"
        case .invalidResponse:
            return "Invalid response from server"
        case .serverError(let code):
            return "Server error (\(code))"
        case .httpError(let code):
            return "Request failed (\(code))"
        }
    }
}

// MARK: - Response Types

struct BatchUploadResult: Decodable {
    let success: Bool
    let inserted: Int
    let total: Int
    let errors: [BatchError]?

    struct BatchError: Decodable {
        let index: Int
        let error: String
    }
}
